src/main/java/io/github/deansg/jeocodio/GeocodioClient.java

```diff
@@ -132,28 +157,57 @@ private String formatFieldsParam(List<String> fields) {
     }
 
     private <T> CompletableFuture<T> sendAsync(HttpRequest httpRequest, Class<T> clazz) {
-        return httpClient.sendAsync(httpRequest, HttpResponse.BodyHandlers.ofString())
-                .thenApply(resp -> {
-                    var body = resp.body();
-                    if (resp.statusCode() != 200) {
-                        throw new GeocodioStatusCodeException(resp.statusCode(), body);
-                    }
-                    return body;
-                })
+        return httpClient.sendAsync(httpRequest, HttpResponse.BodyHandlers.ofInputStream())
+                .thenApply(this::readResponse)
                 .thenApply(str -> this.gson.fromJson(str, clazz));
     }
 
+    private String readResponse(HttpResponse<InputStream> resp) {
+        try (var inputStream = getResponseInputStream(resp)) {
+            String json = new String(inputStream.readAllBytes(), StandardCharsets.UTF_8);
+            if (resp.statusCode() != 200) {
+                throw new GeocodioStatusCodeException(resp.statusCode(), json);
+            }
+            return json;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private InputStream getResponseInputStream(HttpResponse<InputStream> resp) throws IOException {
+        var inputStream = resp.body();
+        if (resp.headers().firstValue("Content-Encoding").orElse("").equals("gzip")) {
+            inputStream = new GZIPInputStream(inputStream);
```

Comment:
Is this a memory leak? Generally input stream creation needs to be done in a try-with block to handle closing of the streams. My IDE was even complaining about `resp.body()` so I also put that in a try-with resource for my sample. Its possible that is a false positive though.

Reply:
The GZIPInputStream should close the underlying stream, and it is used in a try-with-resources in the calling method. Theoretically if the GZIPInputStream creation itself fails it might somehow cause a leak, but honestly I'm not sure we need to explicitly address that. See for example:
https://stackoverflow.com/questions/53502626/does-java-http-client-handle-compression
If you think we should handle it just in case then I can add another try-catch specifically in this method, and call close on the raw stream.

Reply:
If you think it is safe, then that's fine. I'm probably overly safe with these types of issues since I've been burned in the past.